AX: Defer isolated tree node removals so they are batched with other queued node updates

hoffmanjoshua opened the pull request:
#### 5bc669a03a79ddd5cf785f2b4308d43a2f76e94b
<pre>
AX: Defer isolated tree node removals so they are batched with other queued node updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=288001">https://bugs.webkit.org/show_bug.cgi?id=288001</a>
<a href="https://rdar.apple.com/145164516">rdar://145164516</a>

Reviewed by Tyler Wilcock.

In 271907@main, we added a new mechanism to de-duplicate node changes by batching them and
processing them in batches off of a timer. This included most types of changes, but did not
include node removals. This means that nodes could possibly be removed before other types
of updates happen, causing the tree to be temporarily out of sync.

To fix this, let&apos;s start to queue node changes as well in `queueNodeRemoval`. This changes
handles that and updates `removeNode` to work off of AXIDs instead of object references
which it has before.

Many layout tests were updated to be asynchronous, or to be made more asynchronous, due to
behavior this change exposed.

* LayoutTests/accessibility/accessibility-object-detached-expected.txt:
* LayoutTests/accessibility/accessibility-object-detached.html:
* LayoutTests/accessibility/canvas-fallback-content-expected.txt:
* LayoutTests/accessibility/canvas-fallback-content.html:
* LayoutTests/accessibility/datetime/input-date-field-labels-and-value-changes.html:
* LayoutTests/accessibility/mac/element-is-ignored-expected.txt:
* LayoutTests/accessibility/mac/element-is-ignored.html:
* LayoutTests/accessibility/mac/focus-moves-cursor-expected.txt:
* LayoutTests/accessibility/mac/focus-moves-cursor.html:
* LayoutTests/accessibility/mac/slider-thumb-value-crash-expected.txt:
* LayoutTests/accessibility/mac/slider-thumb-value-crash.html:
* LayoutTests/accessibility/mac/stale-textmarker-crash.html:
* LayoutTests/accessibility/mac/textmarker-routines-expected.txt:
* LayoutTests/accessibility/mac/textmarker-routines.html:
* LayoutTests/accessibility/title-ui-element-correctness.html:
* LayoutTests/platform/mac/accessibility/canvas-fallback-content-expected.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::removeNode):
(WebCore::AXIsolatedTree::removeSubtreeFromNodeMap):
(WebCore::AXIsolatedTree::queueNodeRemoval):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/290812@main">https://commits.webkit.org/290812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af62ae261ae91e745dda0cf1839fe3dc22efc78b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/133 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96104 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41865 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70009 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27538 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82543 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50335 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8187 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/109 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40995 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78509 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/126 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98085 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18302 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79017 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18561 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78368 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78218 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19347 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22729 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/80 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23635 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18029 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21490 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19808 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1182,27 +1182,18 @@ void AXIsolatedTree::updateRootScreenRelativePosition()
 }
 
-void AXIsolatedTree::removeNode(const AccessibilityObject& axObject)
+void AXIsolatedTree::removeNode(AXID nodeID, std::optional<AXID> parentID)
```
Small thing, but what do you think about calling this `AXID axID` rather than `nodeID`? I personally don't like the "node" naming scheme we've adopted through `AXIsolatedTree` since DOM nodes are also "nodes" (and canonically, the `Node` class). Have been meaning to change it, but it's low on my list.

hoffmanjoshua replied:
yep, can easily change this!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1219,9 +1210,8 @@ void AXIsolatedTree::removeSubtreeFromNodeMap(std::optional<AXID> objectID, Acce
-    auto axParentID = axParent ? std::optional { axParent->objectID() } : std::nullopt;
     auto actualParentID = m_nodeMap.get(*objectID).parentID;
-    if (axParentID != actualParentID) {
+    if (!axParentID || *axParentID != actualParentID) {
```
Adding a `!axParentID` check means we'll never delete the root node. Any reason why we can't continue comparing `axParentID != actualParentID`?

hoffmanjoshua replied:
It should be okay to not have this actually—I think if the parentID is null and it actually should have a parent the other condition will handle this anyways.

hoffmanjoshua replied:
It turns out we need this condition so we aren't dereferencing `axParentID` without null checking. But, what we really need is to check:
1. If the two values differ in whether they are nullable.
2. If both values are null

We want to _not_ break in case (2), but break in case (1).

I've updated this condition + added a comment to reflect that.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1411,6 +1401,27 @@ void AXIsolatedTree::queueNodeUpdate(AXID objectID, const NodeUpdateOptions& opt
+{
+    ASSERT(isMainThread());
+
+    auto labeledObjectIDs = axObjectCache() ? axObjectCache()->relatedObjectIDsFor(axObject, AXRelationType::LabelFor, AXObjectCache::UpdateRelations::No) : std::nullopt;
```
For optionals, I like to specify the type `std::optional` rather than `auto`, since it's not obvious that this function returns a `std::optional` based on the name in my opinion. What do you think?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -406,9 +406,9 @@ class AXIsolatedTree : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AX
     bool isUnconnectedNode(std::optional<AXID> axID) const { return axID && m_unconnectedNodes.contains(*axID); }
     // Removes the corresponding isolated object and all descendants from the m_nodeMap and queues their removal from the tree.
-    void removeNode(const AccessibilityObject&);
+    void removeNode(AXID, std::optional<AXID>);
```
It could be nice to have a comment explaining what the second parameter is, e.g.:

```cpp
void removeNode(AXID, std::optional<AXID> /* parentID */);
```

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -406,9 +406,9 @@ class AXIsolatedTree : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AX
+    void removeNode(AXID, std::optional<AXID>);
     // Removes the given node and all its descendants from m_nodeMap.
-    void removeSubtreeFromNodeMap(std::optional<AXID>, AccessibilityObject*);
+    void removeSubtreeFromNodeMap(std::optional<AXID>, std::optional<AXID>);
```
Same here.

twilco commented on LayoutTests/accessibility/datetime/input-date-field-labels-and-value-changes.html:
```diff
@@ -31,6 +31,7 @@
 
     output+= "Original subfield values:\n";
     var subfields = getAndLogSubfields(axDateInput);
+    var yearValue;
```
Does this new variable actually get used?

hoffmanjoshua replied:
Ah nope, this is leftover. Will remove this

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1219,9 +1210,10 @@ void AXIsolatedTree::removeSubtreeFromNodeMap(std::optional<AXID> objectID, Acce
+    Markable<AXID> actualParentID = m_nodeMap.get(*objectID).parentID;
+    // If the axParentID and actualParentID differ in whether they are null, or if the values differ, break early.
+    // If both are null, we are likely at the parent.
+    if ((bool)axParentID != (bool)actualParentID || (axParentID && actualParentID && *axParentID != actualParentID)) {
```
Is it possible to use `static_cast<bool>` instead of `(bool)`? I think the former is safer.

twilco replied:
```
(axParentID && actualParentID && *axParentID != actualParentID)
```

Do we need to dereference `actualParentID` too?

Can't we just check whether one or the other is truthy? Since our prior condition will return if they have different bool states?

hoffmanjoshua replied:
Hmm, we can likely just not dereference `axParentID`, and just compare the optional and marked values (which should be able to resolve themselves). Will test this out before updating.

twilco reviewed (approved):
r+ with these last comments

hoffmanjoshua committed:
AX: Defer isolated tree node removals so they are batched with other queued node updates
https://bugs.webkit.org/show_bug.cgi?id=288001
rdar://145164516

Reviewed by Tyler Wilcock.

In 271907@main, we added a new mechanism to de-duplicate node changes by batching them and
processing them in batches off of a timer. This included most types of changes, but did not
include node removals. This means that nodes could possibly be removed before other types
of updates happen, causing the tree to be temporarily out of sync.

To fix this, let's start to queue node changes as well in `queueNodeRemoval`. This changes
handles that and updates `removeNode` to work off of AXIDs instead of object references
which it has before.

Many layout tests were updated to be asynchronous, or to be made more asynchronous, due to
behavior this change exposed.

* LayoutTests/accessibility/accessibility-object-detached-expected.txt:
* LayoutTests/accessibility/accessibility-object-detached.html:
* LayoutTests/accessibility/canvas-fallback-content-expected.txt:
* LayoutTests/accessibility/canvas-fallback-content.html:
* LayoutTests/accessibility/datetime/input-date-field-labels-and-value-changes.html:
* LayoutTests/accessibility/mac/element-is-ignored-expected.txt:
* LayoutTests/accessibility/mac/element-is-ignored.html:
* LayoutTests/accessibility/mac/focus-moves-cursor-expected.txt:
* LayoutTests/accessibility/mac/focus-moves-cursor.html:
* LayoutTests/accessibility/mac/slider-thumb-value-crash-expected.txt:
* LayoutTests/accessibility/mac/slider-thumb-value-crash.html:
* LayoutTests/accessibility/mac/stale-textmarker-crash.html:
* LayoutTests/accessibility/mac/textmarker-routines-expected.txt:
* LayoutTests/accessibility/mac/textmarker-routines.html:
* LayoutTests/accessibility/title-ui-element-correctness.html:
* LayoutTests/platform/mac/accessibility/canvas-fallback-content-expected.txt:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::remove):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateChildren):
(WebCore::AXIsolatedTree::removeNode):
(WebCore::AXIsolatedTree::removeSubtreeFromNodeMap):
(WebCore::AXIsolatedTree::queueNodeRemoval):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/290812@main
